martha/martha_v3.js

```diff
@@ -534,12 +552,70 @@ async function marthaV3Handler(req, res) {
         try {
             bondSA = await apiAdapter.getJsonFrom(bondSAKeyUrl, auth);
         } catch (error) {
-            logAndSendServerError(res, error, 'Received error contacting Bond.');
-            return;
+            throw new RemoteServerError(error, 'Received error contacting Bond.');
         }
     }
 
-    sendResponse(res, requestedFields, drsResponse, fileName, bondProvider, bondSA, accessUrl);
+    Object.assign(params, {
+        drsResponse,
+        fileName,
+        accessUrl,
+        bondSA,
+    });
+}
+
+function buildResponseInfo(params) {
+    const {
+        requestedFields,
+        bondProvider,
+        drsResponse,
+        fileName,
+        accessUrl,
+        bondSA,
+    } = params;
+
+    const fullResponse =
+        requestedFields.length
+            ? convertToMarthaV3Response(drsResponse, fileName, bondProvider, bondSA, accessUrl)
+            : {};
+    const partialResponse = mask(fullResponse, requestedFields.join(","));
+
+    Object.assign(params, {
+        partialResponse,
+    });
+}
+
+async function marthaV3Handler(req, res) {
+    try {
+        // This function counts on the request posting data as "application/json" content-type.
+        // See: https://cloud.google.com/functions/docs/writing/http#parsing_http_requests for more details
+        const {url, fields: requestedFields = MARTHA_V3_DEFAULT_FIELDS} = (req && req.body) || {};
+        const {authorization: auth, 'user-agent': userAgent} = req.headers;
+        const ip = req.ip;
+
+        const params = {
+            url,
+            requestedFields,
+            auth,
+            userAgent,
+            ip,
+        };
+
+        buildRequestInfo(params);
+        await retrieveFromServers(params);
+        buildResponseInfo(params);
+
+        res.status(200).send(params.partialResponse);
+    } catch (error) {
+        if (error instanceof BadRequestError) {
+            logAndSendBadRequest(res, error.cause);
+        } else if (error instanceof RemoteServerError) {
+            logAndSendServerError(res, error.cause, error.description);
+        } else {
+            console.error(`Uncaught error: ${error}`);
+            throw error;
```

Comment:
Since we chatted about this a little in the mob, I remembered that letting an exception escape from a cloud function causes GCP to destroy that instance instead of letting it continue handling requests. Since we don't know what went wrong by the time we get here, killing the instance is probably a good idea. 👍